component/metrics/scraper/appendable.go

```diff
@@ -0,0 +1,86 @@
+package scraper
+
+import (
+	"context"
+	"sync"
+
+	"github.com/prometheus/prometheus/model/labels"
+
+	"github.com/grafana/agent/component/metrics"
+	"github.com/prometheus/prometheus/model/exemplar"
+	"github.com/prometheus/prometheus/storage"
+)
+
+type scrapeAppendable struct {
+	mut       sync.Mutex
+	buffer    map[int64][]*metrics.FlowMetric
+	receivers []*metrics.Receiver
+}
+
+func newScrapeAppendable(receiver []*metrics.Receiver) *scrapeAppendable {
+	return &scrapeAppendable{
+		buffer:    make(map[int64][]*metrics.FlowMetric),
+		receivers: receiver,
+	}
+}
+
+func (s *scrapeAppendable) Append(ref storage.SeriesRef, l labels.Labels, t int64, v float64) (storage.SeriesRef, error) {
+	// BIG TODO is that we probably want to move refid creation and caching into a shared cache between wal and scraper at some point.
+	// 	in the below the refid is never cached by the scrape pool which makes it sad. Its likely this / refid caching should
+	//  should create the refid cache
+	s.mut.Lock()
+	defer s.mut.Unlock()
+	if len(s.receivers) == 0 {
+		return 0, nil
+	}
+	_, found := s.buffer[t]
+	if !found {
+		set := make([]*metrics.FlowMetric, 0)
+		s.buffer[t] = set
+	}
+	// If ref is 0 then lets grab a global id
+	if ref == 0 {
+		ref = storage.SeriesRef(metrics.GlobalRefMapping.CreateGlobalRefID(l))
+	}
+	s.buffer[t] = append(s.buffer[t], &metrics.FlowMetric{
```

Comment:
What do we win/lose by batching metrics together by timestamp? 

Reading through Prometheus' scrape loop code, Append() will be called by the same timestamp for every scrape iteration (rounded to avoid jitter) so I _think_ that this buffer map would end up having as many entries as scrapers.

Reply:
Number of function calls is probably the primary win. If you have a scraper -> remote write there isn't any value since they both get converted to single calls. But adding filters in between would increase the overhead. IE without batching if you had 10 metrics then scraper -> remote write would be 20 total calls , but if you had two filters in between it would be 40, whereas with batching it would 23. 

The loss is some possible memory usage but I think that is a wash. I should prove that though.

Reply:
Sorry, I didn't get the point about the filter. 😕  

I'm trying to understand what would change if the buffer would be a `[]metrics.FlowMetric` and it was all batched into a single call to Receive().